lib/Sema/SemaExpr.cpp

```diff
@@ -12490,36 +12495,74 @@ ExprResult Sema::ActOnBoundsCastExpr(Scope *S, SourceLocation LParenLoc,
     }
   }
 
-  return BuildBoundsCastExpr(LParenLoc, castTInfo, RParenLoc, E1,
-                             dyn_cast<BoundsExpr>(bounds.get()), kind);
+  return BuildBoundsCastExpr(OpLoc, Kind, castTInfo,
+                             SourceRange(LAngleBracketLoc, RAngleBracketLoc),
+                             SourceRange(LParenLoc, RParenLoc), E1,
+                             dyn_cast<BoundsExpr>(bounds.get()));
 }
 
-ExprResult Sema::GenerateBoundsExpr(Expr *E2, Expr *E3) {
+ExprResult Sema::GenerateBoundsExpr(Expr *E1, Expr *E2, Expr *E3,
+                                    QualType DestTy, int SyntaxType,
+                                    tok::TokenKind Kind,
+                                    SourceLocation TypeLoc) {
   ExprResult Result(true);
   ExprResult ResE2 = CorrectDelayedTyposInExpr(E2);
   ExprResult ResE3 = CorrectDelayedTyposInExpr(E3);
 
-  if (!ResE2.isUsable() && !ResE3.isUsable()) {
-    llvm::APInt I = llvm::APInt(1, 1, false);
-    uint64_t Bits = I.getZExtValue();
-    unsigned Width = Context.getIntWidth(Context.UnsignedLongLongTy);
-    llvm::APInt ResultVal(Width, Bits);
-    IntegerLiteral *One = IntegerLiteral::Create(
-        Context, ResultVal, Context.UnsignedLongLongTy, SourceLocation());
-    Result =
-        ActOnCountBoundsExpr(SourceLocation(), BoundsExpr::Kind::ElementCount,
-                             One, SourceLocation());
-  } else if (ResE2.isUsable() && !ResE3.isUsable()) {
-    Result =
-        ActOnCountBoundsExpr(SourceLocation(), BoundsExpr::Kind::ElementCount,
-                             ResE2.get(), SourceLocation());
-  } else if (ResE2.isUsable() && ResE3.isUsable()) {
-    Result = ActOnRangeBoundsExpr(SourceLocation(), ResE2.get(), ResE3.get(),
-                                  SourceLocation());
+  switch (SyntaxType) {
+  case BoundsCastExpr::SyntaxType::Single:
+    if (DestTy->isCheckedPointerPtrType() || DestTy->isUncheckedPointerType()) {
+      llvm::APInt I = llvm::APInt(1, 1, false);
+      uint64_t Bits = I.getZExtValue();
+      unsigned Width = Context.getIntWidth(Context.UnsignedLongLongTy);
+      llvm::APInt ResultVal(Width, Bits);
+      IntegerLiteral *One = IntegerLiteral::Create(
+          Context, ResultVal, Context.UnsignedLongLongTy, SourceLocation());
+      Result =
+          ActOnCountBoundsExpr(SourceLocation(), BoundsExpr::Kind::ElementCount,
+                               One, SourceLocation());
+    } else if (DestTy->isCheckedPointerArrayType()) {
+      Diag(TypeLoc,
+           diag::err_bounds_cast_array_pointer_type_with_single_syntax)
+          << DestTy;
+      Result = CreateInvalidBoundsExpr();
+    } else if (DestTy->isIntegralType(Context)) {
+      Diag(TypeLoc, diag::err_bounds_cast_integral_with_single_syntax)
+          << DestTy;
+      Result = CreateInvalidBoundsExpr();
+    } else
```

Comment:
It would be incorrect to fall into this case and not produce an error message.   The compiler would generate code in that case, even though the bounds would be invalid.

If you change the `else` clause that is above as I suggest, this is no longer an issue.